Add logs to CPVM connection process

bernardodemarco opened the pull request:
### Description

The logs related to the CPVM connection are not enough to perform a great troubleshooting process.

This PR, therefore, adds logs and improves the existing ones in order to facilitate the CPVM connection troubleshooting. Furthermore, the legacy CPVM log proxy classes have been removed, in order to let the CPVM classes use log4j dependencies directly.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### Screenshots (if appropriate):

### How Has This Been Tested?

- I created a VM and accessed it through a CPVM.
- Meanwhile, I verified that the messages were logged as expected.

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:
```diff
@@ -155,31 +157,35 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
     @OnWebSocketError
     public void onError(Throwable cause) {
-        s_logger.error("Error on websocket", cause);
+        s_logger.error("Error on WebSocket.", cause);
```
```suggestion
        s_logger.error("Error on closing WebSocket.");
        s_logger.debug("Error on WebSocket.", cause);
```

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java:
```diff
@@ -72,9 +72,9 @@ public boolean isHostConnected() {
-            s_logger.info("Front end has been idle for too long");
+        long unusedTime = System.currentTimeMillis() - getClientLastFrontEndActivityTime();
+        if (!connectionAlive || unusedTime > ConsoleProxy.VIEWER_LINGER_SECONDS * 1000) {
+            s_logger.info("Front end has been idle for too long ({}ms).", unusedTime);
```
```suggestion
            s_logger.info("Front end has been idle for too long ({} ms).", unusedTime);
```

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java:
```diff
@@ -155,7 +156,7 @@ private boolean sendReadBytesToNoVNC(byte[] b, int readBytes) {
             updateFrontEndActivityTime();
         } catch (WebSocketException | IOException e) {
-            s_logger.debug("Connection exception", e);
+            s_logger.error("VNC server connection exception.", e);
```
stacktrace in `error`

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java:
```diff
@@ -300,8 +301,10 @@ private void connectClientToVNCServer(String tunnelUrl, String tunnelSession, St
+            s_logger.info("Connection to VNC server has been established successfully.");
         } catch (Throwable e) {
-            s_logger.error("Unexpected exception", e);
+            s_logger.error("Unexpected exception while connecting to VNC server.", e);
```
```suggestion
            s_logger.error("Unexpected exception while connecting to VNC server, {}", e.getLocalizedMessage());
            s_logger.debug("Exception while connecting to VNC server.", e);
```

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java:
```diff
@@ -175,8 +176,9 @@ public byte[] authenticateTunnel(String password)
 
-                s_logger.error("Authentication to VNC server is failed. Reason: " + reason);
-                throw new RuntimeException("Authentication to VNC server is failed. Reason: " + reason);
+                String msg = String.format("Authentication to VNC server has failed. Reason: %s", reason);
+                s_logger.error(msg);
```
:+1:

bernardodemarco commented:
@DaanHoogland, thanks for your review!

As logging standards in CloudStack are still under discussion (#8746), I'd rather keep logging the exceptions stacktraces at the `error` level.

Honestly, I do not see the point in turning off the stacktraces at the `error` level. For operators that have a minimum software development background, the stacktraces can be a useful way to perform troubleshooting processes. 

On the other hand, if the operator has no knowledge of development, he might not understand the stacktraces, and that is fine, causing no harm.

Besides that, following along with what you've proposed in #8746, what happens when the operator chooses to turn the stacktraces down by setting the logs level to `error`? Are all exception logs be lost, since they are gonna be at `debug` level? If so, the difficulty of performing troubleshootings would increase significantly, as some errors/problems might be difficult to reproduce. 

Honestly, I'm looking forward to see the outcome of the #8746 discussion and I commit myself to participate on it. However, as I've mentioned before, I'd rather keep the proposed changes related to the stacktraces at the `error` level the way they are.

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:
```diff
@@ -104,20 +105,20 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
             port = Integer.parseInt(portStr);
         } catch (NumberFormatException e) {
-            logger.error("Invalid port value in query string: {}. Expected a number.", portStr, e);
+            logger.error(String.format("Invalid port value in query string: %s. Expected a number.", portStr), e);
```
can you undo this change?

bernardodemarco replied:
I'm not really sure why, but the `logger` fields of `com.cloud.consoleproxy.ConsoleProxyNoVNCHandler` and `com.cloud.consoleproxy.vnc.NoVncClient` are not instantiated directly from `log4j`. They're generated from a static method from the `com.cloud.consoleproxy.util.Logger` helper class.

The problem is that this helper class does not implement the new APIs provided by the new version of `log4j`, especially those that accept multiple parameters. As a consequence of that, `String.format()` must be used to log messages in those two classes.

IMO, it should be evaluated if the helper class is still needed. If so, we need to implement new methods to be able to use the new syntax provided by `log4j2`.

DaanHoogland replied:
I see, That was a good initiative back in 2013, to create a library facade, it had not been implemented fully in the project though;
```
// logger facility for dynamic switch between console logger used in Applet and log4j based logger
```
This might be an oversight in the log4j migration. `com.cloud.consoleproxy.ConsoleProxyLoggerFactory` is still using log4j v1 classes under the hood.

I personally think we should have implemented such a proxy library, for the whole project and implemented log4j under the hood. I think for now we will have to do with 'fixing' the console proxy classes.

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:
```diff
@@ -104,20 +105,20 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
                 ajaxSessionId = Long.parseLong(ajaxSessionIdStr);
             } catch (NumberFormatException e) {
-                logger.error("Invalid ajaxSessionId (sess) value in query string: {}. Expected a number.", ajaxSessionIdStr, e);
+                logger.error(String.format("Invalid ajaxSessionId (sess) value in query string: %s. Expected a number.", ajaxSessionIdStr), e);
```
can you undo this change

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:
```diff
@@ -147,7 +150,7 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
             logger.debug("Viewer has been created successfully.");
         } catch (Exception e) {
-            logger.error("Failed to create viewer due to {}", e.getMessage(), e);
+            logger.error(String.format("Failed to create viewer due to %s", e.getMessage()), e);
```
can you undo this change?

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:
```diff
@@ -180,7 +182,7 @@ public void onClose(Session session, int statusCode, String reason) throws IOExc
     @OnWebSocketFrame
     public void onFrame(Frame f) throws IOException {
-        logger.trace("Sending client frame of {} bytes.", f.getPayloadLength());
+        logger.trace(String.format("Sending client frame of %s bytes.", f.getPayloadLength()));
```
can you undo this change?

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java:
```diff
@@ -79,15 +79,15 @@ public void connectTo(String host, int port, String path, String session, boolea
         }
 
-        logger.info("Connecting to VNC server {}:{} ...", host, port);
+        logger.info(String.format("Connecting to VNC server %s:%s ...", host, port));
```
can you undo this change?

DaanHoogland commented on services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java:
```diff
@@ -79,15 +79,15 @@ public void connectTo(String host, int port, String path, String session, boolea
     public void connectTo(String host, int port) {
         // Connect to server
-        logger.info("Connecting to VNC server {}:{} ...", host, port);
+        logger.info(String.format("Connecting to VNC server %s:%s ...", host, port));
```
can you undo this change?

DaanHoogland reviewed (approved):
clgtm, can we now remove `com.cloud.consoleproxy.util.Logger` from the code base @bernardodemarco ? Just to deal with the technical debt. (we can always add a new proxy once we get wise ;)

bernardodemarco commented:
> clgtm, can we now remove `com.cloud.consoleproxy.util.Logger` from the code base @bernardodemarco ? Just to deal with the technical debt. (we can always add a new proxy once we get wise ;)

@DaanHoogland, I have just removed the CPVM log proxy classes and updated the PR's description.

DaanHoogland commented:
> > clgtm, can we now remove `com.cloud.consoleproxy.util.Logger` from the code base @bernardodemarco ? Just to deal with the technical debt. (we can always add a new proxy once we get wise ;)
> 
> @DaanHoogland, I have just removed the CPVM log proxy classes and updated the PR's description.

thanks for honouring my obnoxious request @bernardodemarco .

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 9981

DaanHoogland commented:
@blueorangutan test alma9 kvm-alma9

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (alma9 mgmt + kvm-alma9) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-10475)</b>
Environment: kvm-alma9 (x2), Advanced Networking with Mgmt server a9
Total time taken: 48808 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8924-t10475-kvm-alma9.zip
Smoke tests completed. 134 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

hsato03 reviewed (approved):
LGTM. 

Tested using the web consoles and monitoring the CPVM logs at the same time. The logs are being created correctly.

JoaoJandre commented:
@blueorangutan package

blueorangutan commented:
@JoaoJandre a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 10637

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✖️ debian ✔️ suse15. SL-JID 10651

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-11075)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 55725 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8924-t11075-kvm-ol8.zip
Smoke tests completed. 138 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_05_vmschedule_test_e2e | `Failure` | 362.40 | test_vm_schedule.py
